radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java

```diff
@@ -115,71 +192,155 @@ public void processGetVerticesRequest(GetVerticesRequest request) {
 		this.syncVerticesRPCSender.sendGetVerticesResponse(request, fetched);
 	}
 
-	public void processGetVerticesResponse(GetVerticesResponse response) {
-		log.info("SYNC_VERTICES: Received GetVerticesResponse {}", response);
+	private void rebuildAndSyncQC(SyncState syncState) {
+		log.info("SYNC_STATE: Rebuilding and syncing QC: sync={} curRoot={}", syncState, vertices.get(rootId.get()));
 
-		final Hash syncTo = (Hash) response.getOpaque();
+		// TODO: check if there are any vertices which haven't been local sync processed yet
+		if (requiresCommittedStateSync(syncState)) {
+			syncState.fetched.sort(Comparator.comparing(Vertex::getView));
+			List<Vertex> nonRootVertices = syncState.fetched.stream().skip(1).collect(Collectors.toList());
+			rebuild(syncState.fetched.get(0), syncState.fetched.get(1).getQC(), nonRootVertices);
+		}
+
+		// At this point we are guaranteed to be in sync with the committed state
+
+		if (!addQC(syncState.qc)) {
+			doQCSync(syncState);
+		}
+	}
+
+	public void processCommittedStateSync(CommittedStateSync committedStateSync) {
+		log.info("SYNC_STATE: synced {}", committedStateSync);
+
+		final Hash syncTo = (Hash) committedStateSync.getOpaque();
 		SyncState syncState = syncing.get(syncTo);
-		if (syncState == null) {
-			return;
+		if (syncState != null) {
+			rebuildAndSyncQC(syncState);
 		}
+	}
 
-		final QuorumCertificate qc = syncState.qc;
+	private void processVerticesResponseForCommittedSync(Hash syncTo, SyncState syncState, GetVerticesResponse response) {
+		log.info("SYNC_STATE: Processing vertices {}", syncState);
 
-		if (response.getVertices().isEmpty()) {
-			log.info("GET_VERTICES failed: {}", qc);
-			// failed
-			return;
+		long stateVersion = syncState.committedVertexMetadata.getStateVersion();
+		List<ECPublicKey> signers = Collections.singletonList(syncState.author);
+		syncState.fetched.addAll(response.getVertices());
+
+		if (syncedStateComputer.syncTo(stateVersion, signers, syncTo)) {
+			rebuildAndSyncQC(syncState);
+		} else {
+			syncState.syncStage = SyncStage.SYNC_TO_COMMIT;
```

Comment:
Should this be using the setter for `syncStage`?

Reply:
yes